src/runtime-rs/crates/hypervisor/src/device/device_manager.rs

```diff
@@ -437,29 +457,61 @@ impl DeviceManager {
         Err(anyhow!("ID are exhausted"))
     }
 
-    pub async fn try_update_device(&mut self, update_config: &UpdateDeviceConfig) -> Result<()> {
+    async fn try_update_device(&mut self, update_config: &DeviceConfig) -> Result<()> {
         let device_id = match update_config {
-            UpdateDeviceConfig::ShareFsCfg(config) => {
-                // try to find the device, if found and just return id.
-                if let Some(matched_device_id) = self.find_device(config.source.clone()).await {
-                    matched_device_id
+            DeviceConfig::ShareFsCfg(config) => {
+                // Try to find the sharefs device. If found, just return.
+                if let Some(target_id) = self.find_device(config.shared_path.clone()).await {
+                    // get the device
+                    let target_device = self
+                        .get_device_info(&target_id)
+                        .await
+                        .context("get device failed")?;
+
+                    // It's safe to get the sharefs device from the enum DeviceType.
+                    if let DeviceType::ShareFs(mut device) = target_device {
+                        // update the mount_config.
+                        device.config.mount_config = config.mount_config.clone();
+                        // update the hashmap with the updated device
+                        self.devices
+                            .insert(target_id.clone(), Arc::new(Mutex::new(device)));
+                    }
+                    target_id
                 } else {
                     return Err(anyhow!(
                         "no matching device was found to do the update operation"
                     ));
                 }
             }
+            // TODO for other Device Type
+            _ => {
+                return Err(anyhow!("update device with unsupported device type"));
+            }
         };
 
         // find the device
-        let device = self
-            .devices
-            .get(&device_id)
-            .context("failed to find device")?;
+        let devices = &self.devices.clone();
+        let device = &devices.get(&device_id).context("failed to find device")?;
         let mut device_guard = device.lock().await;
 
-        // handle update
-        device_guard.update(self.hypervisor.as_ref()).await?;
+        // handle update and error
+        if let Err(e) = device_guard.update(self.hypervisor.as_ref()).await {
+            match device_guard.get_device_info().await {
+                DeviceType::ShareFs(mut dev) => {
+                    // rollback the mount_config.
+                    dev.config.mount_config = None;
```

Comment:
Is it guaranteed that the previous value was `Node`?

Reply:
Good question! Thx @pmores 
Fixed

Reply:
I notice that you removed the rollback altogether but frankly what I had in mind was rather, either store the original value so that it can be restored here, or maybe even just work with a clone of the `Device` for the duration of the function and only commit it back to `devices` (by `self.devices.insert()`ing it) once it's clear that everything succeeded?

Reply:
Thanks @Apokleos !